ambry-store/src/main/java/com.github.ambry.store/BlobStoreCompactor.java

```diff
@@ -0,0 +1,864 @@
+            options.close();
+          }
+        }
+      } else if (!srcIndex.isExpired(value)) {
+        // unexpired PUT entry.
+        if (deletesInEffect) {
+          FileSpan deleteSearchSpan = new FileSpan(indexSegmentStartOffset, startOffsetOfLastIndexSegmentForDelete);
+          IndexValue searchedValue = srcIndex.findKey(indexEntry.getKey(), deleteSearchSpan);
+          if (!searchedValue.isFlagSet(IndexValue.Flags.Delete_Index)) {
+            // PUT entry that has not expired and is not considered deleted.
+            validEntries.add(indexEntry);
+          } else {
+            logger.trace("Skipping {} because it is a deleted PUT", indexEntry.getKey());
+          }
+        } else {
+          // valid PUT entry
+          validEntries.add(indexEntry);
+        }
+      } else {
+        logger.trace("Skipping {} because it is an expired PUT", indexEntry.getKey());
+      }
+    }
+    return validEntries;
+  }
+
+  /**
+   * Checks if a record already exists in the target log (swap spaces).
+   * @param key the {@link StoreKey} of the record.
+   * @param srcValue the {@link IndexValue} obtained from the application index.
+   * @return {@code true} if the record already exists in the target log, {@code false} otherwise.
+   * @throws StoreException if there is any problem with using the index.
+   */
+  private boolean alreadyExistsInTgt(StoreKey key, IndexValue srcValue) throws StoreException {
+    IndexValue tgtValue = tgtIndex.findKey(key);
+    return tgtValue != null && srcValue.isFlagSet(IndexValue.Flags.Delete_Index) == tgtValue.isFlagSet(
+        IndexValue.Flags.Delete_Index);
+  }
+
+  /**
+   * Copies the given {@code srcIndexEntries} from the given log segment into the swap spaces.
+   * @param logSegmentToCopy the {@link LogSegment} to copy from.
+   * @param srcIndexEntries the {@link IndexEntry}s to copy.
+   * @param lastModifiedTimeSecs the last modified time of the source index segment.
+   * @return @code true} if all the records  were copied. {@code false} if some records were not copied either because
+   * there was no more capacity or because a shutdown was initiated.
+   * @throws IOException if there were I/O errors during copying.
+   * @throws StoreException if there are any problems reading or writing to store components.
+   */
+  private boolean copyRecords(LogSegment logSegmentToCopy, List<IndexEntry> srcIndexEntries, long lastModifiedTimeSecs)
+      throws IOException, StoreException {
+    boolean copiedAll = true;
+    long totalCapacity = tgtLog.getCapacityInBytes();
+    FileChannel fileChannel = Utils.openChannel(logSegmentToCopy.getView().getFirst(), false);
+    long writtenLastTime = 0;
+    for (IndexEntry srcIndexEntry : srcIndexEntries) {
+      IndexValue srcValue = srcIndexEntry.getValue();
+      long usedCapacity = tgtIndex.getLogUsedCapacity();
+      if (isActive && tgtLog.getCapacityInBytes() - usedCapacity >= srcValue.getSize()) {
```

Comment:
can you add brackets for (tgtLog.getCapacityInBytes() - usedCapacity >= srcValue.getSize())